### PR TITLE
Add gender field

### LINKED ISSUE #34: Add gender
Add gender fields for a person
- [x] show gender of person when listing them
- [x] Allow them to add

### PR DESCRIPTION
**Pull request by athuyaoo**
Added gender field to Hall-y

image

image

Closes #34

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -121,4 +122,21 @@ public static Tag parseTag(String tag) throws ParseException {
         }
         return tagSet;
     }
+
+
+    /**
```
Comment (jieffyy) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:127]: A very minor nit, but is there an additional linespace between the previous method and this method?

File: src/main/java/seedu/address/model/person/Gender.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Resident's gender in Hall-y.
+ * Guarantees: immutable; is valid as declared in
+ */
+public class Gender {
+    public static final String MESSAGE_CONSTRAINTS = "Gender should either be M for male or F for female.";
+    public final GenderType type;
```
Comment (jieffyy) [inline src/main/java/seedu/address/model/person/Gender.java:13]: Should `GenderType type` be a private variable as stated by the coding standards?

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 3]
```diff
@@ -102,8 +106,17 @@ public Person toModelType() throws IllegalValueException {
         }
         final Address modelAddress = new Address(address);
 
+
```
Comment (jieffyy) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java]: Minor nit again 😟 . Is there an additional linespace between this if block and the previous one?

File: src/test/java/seedu/address/testutil/PersonUtil.java [hunk 4]
```diff
@@ -34,6 +35,7 @@ public static String getPersonDetails(Person person) {
         sb.append(PREFIX_PHONE + person.getPhone().value + " ");
         sb.append(PREFIX_EMAIL + person.getEmail().value + " ");
         sb.append(PREFIX_ADDRESS + person.getAddress().value + " ");
+        sb.append(PREFIX_GENDER + person.getGender().type.getOption() + " ");
         person.getTags().stream().forEach(
             s -> sb.append(PREFIX_TAG + s.tagName + " ")
         );
```
Comment (jieffyy) [inline src/test/java/seedu/address/testutil/PersonUtil.java:42]: Should the `getEditPersonDescriptorDetails()` in the file be updated also?

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 5]
```diff
@@ -45,40 +47,43 @@
 
     @Test
     public void parse_allFieldsPresent_success() {
-        Person expectedPerson = new PersonBuilder(BOB).withTags(VALID_TAG_FRIEND).build();
+        Person expectedPerson = new PersonBuilder(BOB).withTags(VALID_TAG_FRIEND).withGender("M").build();
```
Comment (jieffyy) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Should this be `GENDER_DESC_BOB` instead?

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 6]
```diff
@@ -45,40 +47,43 @@
...
+                + ADDRESS_DESC_BOB + GENDER_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
 
         // multiple addresses - last address accepted
         assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_AMY
-                + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
+                + ADDRESS_DESC_BOB + GENDER_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
 
         // multiple tags - all accepted
-        Person expectedPersonMultipleTags = new PersonBuilder(BOB).withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND)
+        Person expectedPersonMultipleTags = new PersonBuilder(BOB).withGender("M")
```
Comment (jieffyy) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Should the `"M"` be `GENDER_DESC_BOB` instead?

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 7]
```diff
@@ -110,32 +119,38 @@ public void parse_compulsoryFieldMissing_failure() {
...
-                + TAG_DESC_HUSBAND + TAG_DESC_FRIEND, Address.MESSAGE_CONSTRAINTS);
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                + INVALID_ADDRESS_DESC + GENDER_DESC_BOB + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
+                Address.MESSAGE_CONSTRAINTS);
 
         // invalid tag
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                + ADDRESS_DESC_BOB + GENDER_DESC_BOB
                 + INVALID_TAG_DESC + VALID_TAG_FRIEND, Tag.MESSAGE_CONSTRAINTS);
```
Comment (jieffyy) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:187]: Would you consider adding a similar test for invalid gender as well?

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 8]
```diff
@@ -97,9 +98,10 @@ private static Person createEditedPerson(Person personToEdit, EditPersonDescript
         Phone updatedPhone = editPersonDescriptor.getPhone().orElse(personToEdit.getPhone());
         Email updatedEmail = editPersonDescriptor.getEmail().orElse(personToEdit.getEmail());
         Address updatedAddress = editPersonDescriptor.getAddress().orElse(personToEdit.getAddress());
+        Gender updatedGender = editPersonDescriptor.getGender().orElse(personToEdit.getGender());
```
Comment (jieffyy) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:106]: Would you consider updating the `MESSAGE_USAGE` in this class as well?

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java [hunk 9]
```diff
@@ -71,6 +72,16 @@ public EditPersonDescriptorBuilder withAddress(String address) {
         return this;
     }
 
+
```
Comment (jieffyy) [inline src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: Should the constructor of this class be updated as well?

**Review by jieffyy — COMMENTED** [review]
Nice work! 👍 

Some areas you can consider updating as well:

Under `EditCommandParserTest.java`, there some methods like `parse_oneFieldSpecified_success()` that could be updated to include parsing just the gender field.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 1 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:127]: Thanks, will change.

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java (hunk 3 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/storage/JsonAdaptedPerson.java]: Sure, thanks.

File: src/main/java/seedu/address/model/person/Gender.java (hunk 2 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/model/person/Gender.java:13]: No, so far all the model properties (Address, Email, Name, etc) all has public final variables to store their values.

File: src/test/java/seedu/address/testutil/PersonUtil.java (hunk 4 above)
Comment (athuyaoo) [reply src/test/java/seedu/address/testutil/PersonUtil.java:42]: Fixed

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 6 above)
Comment (athuyaoo) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java]: Thanks, Fixed

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java (hunk 7 above)
Comment (athuyaoo) [reply src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:187]: Sure, done

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 8 above)
Comment (athuyaoo) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:106]: Done

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 10]
```diff
@@ -45,13 +48,16 @@
...
+    public static final String GENDER_DESC_AMY = " " + PREFIX_GENDER + VALID_GENDER_AMY;
+    public static final String GENDER_DESC_BOB = " " + PREFIX_GENDER + VALID_GENDER_BOB;
     public static final String TAG_DESC_FRIEND = " " + PREFIX_TAG + VALID_TAG_FRIEND;
     public static final String TAG_DESC_HUSBAND = " " + PREFIX_TAG + VALID_TAG_HUSBAND;
 
     public static final String INVALID_NAME_DESC = " " + PREFIX_NAME + "James&"; // '&' not allowed in names
     public static final String INVALID_PHONE_DESC = " " + PREFIX_PHONE + "911a"; // 'a' not allowed in phones
     public static final String INVALID_EMAIL_DESC = " " + PREFIX_EMAIL + "bob!yahoo"; // missing '@' symbol
     public static final String INVALID_ADDRESS_DESC = " " + PREFIX_ADDRESS; // empty string not allowed for addresses
+    public static final String INVALID_GENDER = " " + PREFIX_GENDER + "A"; // '*' not allowed in tags
```
Comment (jieffyy) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:67]: Minor typo here in the comments.

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java (hunk 9 above)
Comment (athuyaoo) [reply src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: Done

**Review by jieffyy — APPROVED** [review]
LGTM! Nice work! 👍

**Pull request merged**